tutorial for adding radius to existing application (#870)
radius-project/docs: docs/content/tutorials/tutorial-add-radius/index.md

---
type: docs
title: "Tutorial: Add Radius to an existing application"
linkTitle: "Existing applications"
description: "Learn how to incrementally add Radius to an existing application"
weight: 600
slug: "add-radius"
categories: "Tutorial"

> tags : ["incremental","adoption"]

Reviewer: I hope we don't use these as tags. Who would click on them?

Author: good point, I will change this.

> tags : ["incremental","adoption"]

Author: ```suggestion
tags : ["kubernetes","kubectl"]
```

---

This tutorial will teach you about adding Radius to an existing application using the built-in Kubernetes integration functionality. You will learn how to:

1. Deploy an existing Kubernetes application
1. Run and test the application
1. Add Radius to the deployed application
1. Confirm that Radius was added and can detect the application

By the end of this tutorial, you will have deployed an existing Kubernetes application and then updated it to add Radius functionality without changing the application or its original Kubernetes configurations.

## Prerequisites

- [Kubernetes cluster]({{< ref "guides/operations/kubernetes/overview#supported-kubernetes-clusters" >}})
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
- [rad CLI]({{< ref howto-rad-cli >}})

## Overview

> As a part of this tutorial you will deploy an existing and containerized Guestbook application originally authored by the Kubernetes community for use in their own tutorials, then add Radius to the deployed application. To incrementally add Radius to an existing application, you will leverage the built-in Kubernetes integration functionality by adding an annotation to the application's Kubernetes deployment manifest. This approach is particularly useful for adding Radius capabilities to applications that are already deployed.

Reviewer 2: ```suggestion
As a part of this tutorial you will deploy an existing containerized Guestbook application originally authored by the Kubernetes community for use in their own tutorials. You will then add Radius to the deployed application. To incrementally add Radius to an existing application, you will leverage the built-in Kubernetes integration functionality by adding an annotation to the application's Kubernetes deployment manifest. This approach is particularly useful for adding Radius capabilities to applications that are already deployed.
```

The Guestbook application consists of a web front end along with primary and secondary Redis containers for storage, all deployed with Kubernetes. For more information about the application and access its source code, see the [Kubernetes tutorial](https://kubernetes.io/docs/tutorials/stateless-application/guestbook/) and their [examples repo](https://github.com/kubernetes/examples/tree/master/guestbook).

> 

Reviewer: I would be good to have a diagram here if possible. We've consistently gotten feedback on tutorials that the diagram is needed.

Reviewer: (non-blocking for this PR)

   <img src="./guestbook-ui.png" alt="Guestbook application" width=500>

   Terminate the port forwarding process by pressing `CTRL+C` in your terminal.

5. Confirm that Radius has not yet been added:

   Run this command to view the state of your application using Radius:

   ```bash
   rad app connections -a demo
   ```

   Since you have not yet added Radius to the application, the `rad` CLI will not recognize your `demo` application and you should see a message in the output similar to:

   ```
   Application "demo" does not exist or has been deleted.
   ```

## Step 3: Add Radius to the Guestbook application

You will now add Radius to the Guestbook application's Kubernetes deployment manifests by making edits to the YAML files in the `deploy` directory. 

1. In each of the YAML files, add the `annotations` property to `metadata`, and then add the `rad.app/enabled: 'true'` annotation. Note that the `'true'` must be surrounded in quotes.

   ```yaml
   ...
   metadata:
      ...
      annotations:
         radapp.io/enabled: 'true'
      ...
   spec:
   ...
   ```

   You should add the annotation to the following files:
      - `deploy/frontend-deployment.yaml`

>       - `deploy/frontend-service.yaml`

Reviewer: This doesn't do anything for the services.

Author: Good point, then it probably makes sense to exclude `Services` from the set of items to modify. I will make the change.

   ```
   NAME            READY   UP-TO-DATE   AVAILABLE   AGE
   redis-master    1/1     1            1           10m
   redis-replica   2/2     2            2           10m
   frontend        3/3     3            3           10m
   NAME            TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)        AGE
   frontend        NodePort    10.43.49.106    <none>        80:30972/TCP   10m
   redis-master    ClusterIP   10.43.233.148   <none>        6379/TCP       10m
   redis-replica   ClusterIP   10.43.118.183   <none>        6379/TCP       10m
   ```

3. Validate that the application is running as expected:

   Once again, run this command to port forward the Guestbook application to port `8080` on your local machine:

   ```bash
   kubectl port-forward -n demo svc/frontend 8080:80
   ```

   Open a browser and navigate to [`http://localhost:8080`](http://localhost:8080). You should see the Guestbook application running:

   <img src="./guestbook-ui.png" alt="Guestbook application" width=500>

   Terminate the port forwarding process by pressing `CTRL+C` in your terminal.

## Step 5: Use Radius to display the state of your application

> Now that Radius has been enabled for your application, run this command once again:

Reviewer 2: ```suggestion
Now that Radius has been enabled for your application, run this command again:
```

```bash
rad app connections -a demo
```

> You should see now see some output containing information and status of your newly Radified applicaiton, similar to:

Reviewer 2: ```suggestion
You should see output containing information and status of your newly "radified" application, similar to:
```
